ssl/statem/statem_srvr.c

```diff
@@ -2686,12 +2717,30 @@ int tls_construct_server_key_exchange(SSL *s, WPACKET *pkt)
 int tls_construct_certificate_request(SSL *s, WPACKET *pkt)
 {
     if (SSL_IS_TLS13(s)) {
-        /* TODO(TLS1.3) for now send empty request context */
-        if (!WPACKET_put_bytes_u8(pkt, 0)) {
-            SSLfatal(s, SSL_AD_INTERNAL_ERROR,
-                     SSL_F_TLS_CONSTRUCT_CERTIFICATE_REQUEST,
-                     ERR_R_INTERNAL_ERROR);
-            return 0;
+        /* Send random context when doing post-handshake auth */
+        if (s->post_handshake_auth == SSL_PHA_REQUEST_PENDING) {
+            OPENSSL_free(s->pha_context);
+            s->pha_context_len = 32;
+            if ((s->pha_context = OPENSSL_malloc(s->pha_context_len)) == NULL
+                    || ssl_randbytes(s, s->pha_context, s->pha_context_len) <= 0
+                    || !WPACKET_sub_memcpy_u8(pkt, s->pha_context, s->pha_context_len)) {
+                SSLfatal(s, SSL_AD_INTERNAL_ERROR,
+                         SSL_F_TLS_CONSTRUCT_CERTIFICATE_REQUEST,
+                         ERR_R_INTERNAL_ERROR);
+                return 0;
+            }
+            /* reset the handshake hash back to just after the ClientFinished */
+            if (!tls13_restore_handshake_digest_for_pha(s)) {
```

Comment:
Yup; not easy to (unit) test this. But, perhaps this means that the KeyUpdate and NewSessionTicket should not be placed into the messaging digest.

Reply:
> Yup; not easy to (unit) test this.

It could be done. Create a connection using `create_ssl_connection()`. Call `SSL_verify_client_post_handshake()` followed by `SSL_do_handshake()` on the server side. On the client side call SSL_key_update() and SSL_do_handshake() without first reading anything from the connection. Then attempt to read on the client side so it sees and processes the CertificateRequest message.

> But, perhaps this means that the KeyUpdate and NewSessionTicket should not be placed into the messaging digest.

Or perhaps we always restore the message digest state at the beginning of a post-handshake sequence

Reply:
If the server receives KeyUpdate between sending the CertRequest and receiving the Certificate... messages (as suggested), it's digest state will be incorrect. So, regardless of when the digest state is restored (which I agree, should be at the latest possible moment), I think KeyUpdate and NewSessionTicket should be excluded from the digests.

Reply:
Oh...right. The penny just dropped that the CertRequest itself is included in the transcript hash. So yes, you are right.

Reply:
A unit test has been added for this in sslapitest.c